core/src/main/java/org/polypheny/db/plan/AlgCluster.java

```diff
@@ -0,0 +1,176 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.polypheny.db.plan;
+
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.Getter;
+import org.polypheny.db.algebra.AlgNode;
+import org.polypheny.db.algebra.core.CorrelationId;
+import org.polypheny.db.algebra.metadata.AlgMetadataProvider;
+import org.polypheny.db.algebra.metadata.AlgMetadataQuery;
+import org.polypheny.db.algebra.metadata.DefaultAlgMetadataProvider;
+import org.polypheny.db.algebra.metadata.MetadataFactory;
+import org.polypheny.db.algebra.metadata.MetadataFactoryImpl;
+import org.polypheny.db.algebra.type.AlgDataTypeFactory;
+import org.polypheny.db.catalog.snapshot.Snapshot;
+import org.polypheny.db.rex.RexBuilder;
+import org.polypheny.db.schema.trait.ModelTrait;
+
+
+/**
+ * An environment for related algebra expressions during the optimization of a query.
+ */
+public class AlgCluster {
+
+    @Getter
+    private final AlgDataTypeFactory typeFactory;
+    @Getter
+    private final AlgPlanner planner;
+    private final AtomicInteger nextCorrel;
+    @Getter
+    private final Map<CorrelationId, AlgNode> mapCorrelToAlg;
+    @Getter
+    private final RexBuilder rexBuilder;
+    @Getter
+    private AlgMetadataProvider metadataProvider;
+    @Getter
+    private MetadataFactory metadataFactory;
+    private final AlgTraitSet emptyTraitSet;
+    private AlgMetadataQuery mq;
+
+    @Getter
+    private final Snapshot snapshot;
+
+
+    /**
+     * Creates a cluster.
+     * <p>
+     * For use only from {@link #create} and {@link AlgOptQuery}.
+     */
+    private AlgCluster( AlgPlanner planner, AlgDataTypeFactory typeFactory, RexBuilder rexBuilder, AlgTraitSet traitSet, Snapshot snapshot ) {
+        this.nextCorrel = new AtomicInteger( 0 );
+        this.mapCorrelToAlg = new HashMap<>();
+        this.planner = Objects.requireNonNull( planner );
+        this.typeFactory = Objects.requireNonNull( typeFactory );
+        this.rexBuilder = rexBuilder;
+
+        // set up a default alg metadata provider, giving the planner first crack at everything
+        setMetadataProvider( DefaultAlgMetadataProvider.INSTANCE );
+        this.emptyTraitSet = traitSet == null ? planner.emptyTraitSet() : traitSet;
+        assert emptyTraitSet.size() == planner.getAlgTraitDefs().size();
+        this.snapshot = snapshot;
```

Comment:
There should be a common interface for both the statistics and the metadata because they essentially do the same but with a different level of sophisticatedness. How difficult is it to unify this?

Reply:
I agree, but this requires quite some adjustments to make it work, as they us different structures and do it at different point "during" the query process.